src/ardrive.ts

```diff
@@ -39,23 +45,64 @@ export class ArDriveAnonymous extends ArDriveType {
 	}
 }
 
+// TODO: ArDrive should accept App-Name and App-Version tags from constructor?
+const commTipMetaTags: GQLTagInterface[] = [
+	{ name: 'App-Name', value: 'ArDrive-CLI' },
+	{ name: 'App-Version', value: '2.0' },
+	{ name: 'Trx-Type', value: 'Community-Tip' }
+];
+
 export class ArDrive extends ArDriveAnonymous {
-	constructor(protected readonly arFsDao: ArFSDAO) {
+	constructor(
+		private readonly wallet: Wallet,
+		private readonly walletDao: WalletDAO,
+		protected readonly arFsDao: ArFSDAO,
+		private readonly communityOracle: CommunityOracle
+	) {
 		super(arFsDao);
 	}
 
+	// TODO: FS shouldn't be reading the files more than once and doesn't belong in this class
+	getFileSize(filePath: string): Bytes {
+		return fs.statSync(filePath).size;
+	}
+
+	async sendCommunityTip(communityWinstonTip: Winston): Promise<ArFSTipData> {
+		const tokenHolder: ArweaveAddress = await this.communityOracle.selectTokenHolder();
+
+		const communityTipResult = await this.walletDao.sendARToAddress(
+			winstonToAr(+communityWinstonTip),
+			this.wallet,
+			tokenHolder,
+			commTipMetaTags
+		);
+
+		return { txId: communityTipResult.trxID, recipient: tokenHolder, winston: communityTipResult.winston };
+	}
+
 	async uploadPublicFile(
 		parentFolderId: FolderID,
 		filePath: string,
 		destinationFileName?: string
 	): Promise<ArFSResult> {
+		const winstonPrice = await this.walletDao.getWinstonPriceForBytes(this.getFileSize(filePath));
```

Comment:
You shouldn't need to do this. You can get the cost of the data transaction directly from the trx object itself, i.e.:
```
dataTrx.reward
```

Reply:
The idea here is that we should check the price plus the community and confirm a wallet's balance before creating the transaction. When we create the transaction, we will inevitably be writing file's Buffer to it which the arweave library will encode into base64. 

If the wallet doesn't have enough balance for both the transaction AND the community tip, we shouldn't be initiating something so resource intensive.

Creating the transaction in ArFSDao, handing the transaction back up to check the balance, then sending back down to another function that will sign/send the transaction and is one solution. But that solution would not be performant with more than one file upload at a time -- because we'd be holding on to all of these `dataTrx` types in memory.

Reply:
You're right about performance and the changes you made to try and do that ended up leaking the Arweave library's types out into the ArDrive layer which I don't think we want. We need the price estimator classes.